nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/download/DownloadServiceImpl.java

```diff
@@ -216,12 +223,164 @@ public String getDownloadUrl(String repositoryName, String docId, String xpath,
 
     @Override
     public String getDownloadUrl(String storeKey) {
-        return DownloadService.NXBIGBLOB + "/" + storeKey;
+        return NXBIGBLOB + "/" + storeKey;
+    }
+
+    /**
+     * Gets the download path and action of the URL to use to download blobs. For instance, from the path
+     * "nxfile/default/3727ef6b-cf8c-4f27-ab2c-79de0171a2c8/files:files/0/file/image.png", the pair
+     * ("default/3727ef6b-cf8c-4f27-ab2c-79de0171a2c8/files:files/0/file/image.png", "downloadFromDoc") is returned.
+     *
+     * @param path the path of the URL to use to download blobs
+     * @return the pair download path and action
+     * @since 9.1
+     */
+    protected Pair<String, String> getDownloadPathAndAction(String path) {
+        if (path.startsWith("/")) {
+            path = path.substring(1);
+        }
+        int slash = path.indexOf('/');
+        if (slash < 0) {
+            return null;
+        }
+        String type = path.substring(0, slash);
+        String downloadPath = path.substring(slash + 1);
+        switch (type) {
+        case NXDOWNLOADINFO:
+            // used by nxdropout.js
+            return Pair.of(downloadPath, INFO);
+        case NXFILE:
+        case NXBIGFILE:
+            return Pair.of(downloadPath, EVENT_NAME_FROM_DOC);
+        case NXBIGZIPFILE:
+        case NXBIGBLOB:
+            return Pair.of(downloadPath, EVENT_NAME);
+        default:
+            return null;
+        }
+    }
+
+    /**
+     * Gets a document from a repository name and a docId.
+     *
+     * @param repository the repository name
+     * @param docId the document id
+     * @return the DocumentModel
+     * @since 9.1
+     */
+    protected DocumentModel getDownloadDocument(String repository, String docId) {
+        try (CoreSession session = CoreInstance.openCoreSession(repository)) {
+            DocumentRef docRef = new IdRef(docId);
+            if (!session.exists(docRef)) {
+                return null;
+            }
+            return session.getDocument(docRef);
+        }
+    }
+
+    @Override
+    public Blob resolveBlobFromDownloadUrl(String url) {
+        String nuxeoUrl = Framework.getProperty("nuxeo.url");
+        if (!url.startsWith(nuxeoUrl)) {
+            return null;
+        }
+        String path = url.substring(nuxeoUrl.length() + 1);
+        Pair<String, String> pair = getDownloadPathAndAction(path);
+        if (pair == null) {
+            return null;
+        }
+        String downloadPath = pair.getLeft();
```

Comment:
You should first check the action (right part) and `return null` if it's not a download.